src/v/bytes/iobuf.h

```diff
@@ -394,40 +396,31 @@ iobuf::try_copy_append(const char* ptr, size_t size) {
         return;
     }
 
-    if (available_bytes() > 0) {
-        if (_frags.back().is_empty()) {
-            pop_back();
-        } else {
-            _frags.back().trim();
-        }
+    if (unlikely(available_bytes() > 0 && _frags.back().is_empty())) {
+        pop_back();
     }
     append(std::make_unique<fragment>(std::move(b)));
 }
 
 /// appends the contents of buffer; might pack values into existing space
 inline void iobuf::append(iobuf&& o) {
-    for (auto& f : o._frags) {
-        auto fsize = f.size();
-
-        if (unlikely(!fsize)) {
-            continue;
-        }
+    while (!o._frags.empty()) {
+        fragment* f = &o._frags.front();
+        o._frags.pop_front();
 
-        if (try_copy_append(f.get(), fsize)) {
+        auto fsize = f->size();
+        if (!fsize || try_copy_append(f->get(), fsize)) {
+            details::dispose_io_fragment(f);
             continue;
         }
 
-        if (available_bytes() > 0) {
-            if (_frags.back().is_empty()) {
-                pop_back();
-            } else {
-                _frags.back().trim();
-            }
+        if (unlikely(!_frags.empty() && _frags.back().is_empty())) {
```

Comment:
I thought about just moving this check to `iobuf::append(fragment*)` as well. Need to check the callers of `iobuf::append(std::unique_ptr<fragment>)` though.